pkg/receive/limiter_config.go

```diff
@@ -21,37 +24,84 @@ func ParseRootLimitConfig(content []byte) (*RootLimitsConfig, error) {
 	if err := yaml.UnmarshalStrict(content, &root); err != nil {
 		return nil, errors.Wrapf(err, "parsing config YAML file")
 	}
+
+	if root.WriteLimits.GlobalLimits.MetaMonitoringURL != "" {
+		u, err := url.Parse(root.WriteLimits.GlobalLimits.MetaMonitoringURL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "parsing meta-monitoring URL")
+		}
+
+		// url.Parse might pass a URL with only path, so need to check here for scheme and host.
+		// As per docs: https://pkg.go.dev/net/url#Parse.
+		if u.Host == "" || u.Scheme == "" {
+			return nil, errors.Newf("%s is not a valid meta-monitoring URL (scheme: %s,host: %s)", u, u.Scheme, u.Host)
+		}
+		root.WriteLimits.GlobalLimits.metaMonitoringURL = u
+	}
+
+	// Set default query if none specified.
+	if root.WriteLimits.GlobalLimits.MetaMonitoringLimitQuery == "" {
+		root.WriteLimits.GlobalLimits.MetaMonitoringLimitQuery = "sum(prometheus_tsdb_head_series) by (tenant)"
+	}
+
 	return &root, nil
 }
 
-type writeLimitsConfig struct {
+func (r RootLimitsConfig) AreHeadSeriesLimitsConfigured() bool {
+	return r.WriteLimits.GlobalLimits.MetaMonitoringURL != "" && (len(r.WriteLimits.TenantsLimits) != 0 || r.WriteLimits.DefaultLimits.HeadSeriesLimit != 0)
+}
+
+type WriteLimitsConfig struct {
 	// GlobalLimits are limits that are shared across all tenants.
-	GlobalLimits globalLimitsConfig `yaml:"global"`
+	GlobalLimits GlobalLimitsConfig `yaml:"global"`
 	// DefaultLimits are the default limits for tenants without specified limits.
-	DefaultLimits defaultLimitsConfig `yaml:"default"`
+	DefaultLimits DefaultLimitsConfig `yaml:"default"`
 	// TenantsLimits are the limits per tenant.
-	TenantsLimits tenantsWriteLimitsConfig `yaml:"tenants"`
+	TenantsLimits TenantsWriteLimitsConfig `yaml:"tenants"`
```

Comment:
Is there a reason to export these methods? Apart from being used in E2E tests, looks like this is not really necessary. Maybe we could configure the E2E tests with YAML? (Might even be beneficial, since we can get config parsing in the E2E test for "free").

Reply:
Clarified F2F, @saswatamcode is right that we do this in Thanos by standard and it eases configuring things when e.g. testing, demoing etc., resolving.